File: kotlin/app/src/main/java/com/vitalcheck/app/domain/usecase/CreateSymptomEntry.kt
package com.vitalcheck.app.domain.usecase

import com.vitalcheck.app.domain.entity.SymptomEntry
import com.vitalcheck.app.domain.repository.ISymptomRepository

/**
 * Use case: criar um novo registro de sintoma.
 *
 * Responsabilidades:
 * 1. Criar a entidade com validação (via SymptomEntry.create)
 * 2. Persistir via repositório
 * 3. Retornar a entidade criada para atualização otimista da UI
 *
 * A validação ocorre na criação da entidade (fail-fast).
 * Se a descrição for inválida, uma IllegalArgumentException é lançada
 * antes de qualquer IO.
 */
class CreateSymptomEntry(
    private val repository: ISymptomRepository
) {
    suspend operator fun invoke(description: String): SymptomEntry {
        val entry = SymptomEntry.create(description = description)
        repository.save(entry)
        return entry
    }
}

File: react-native/src/domain/usecases/CreateSymptomEntry.ts
import { SymptomEntry, createSymptomEntry } from '../entities/SymptomEntry';
import { ISymptomRepository } from '../repositories/ISymptomRepository';

/**
 * Use Case: Criar um novo registro de sintoma.
 *
 * Coordena a criação da entidade com validação de domínio
 * e sua persistência através do repositório.
 *
 * O ID é gerado externamente (injetado via factory) para manter
 * testabilidade e separação de responsabilidades.
 */
export class CreateSymptomEntry {
  constructor(
    private readonly repository: ISymptomRepository,
    private readonly generateId: () => string
  ) {}

  /**
   * Cria e persiste um novo registro de sintoma.
   *
   * @param description - Descrição textual do sintoma
   * @returns O registro criado
   * @throws Error se a descrição for inválida ou a persistência falhar
   */
  async execute(description: string): Promise<SymptomEntry> {
    const id = this.generateId();
    const entry = createSymptomEntry(id, description, new Date());

    await this.repository.save(entry);

    return entry;
  }
}

